spec/client/ics-028-wasm-client/README.md

```diff
@@ -0,0 +1,451 @@
+  root: MerkleRoot
+  type: String
+}
+```
+
+### Height
+
+The height of a WASM light client instance consists of two `uint64`s: the epoch number, and the height in the epoch.
+
+```typescript
+interface Height {
+  epochNumber: uint64
+  epochHeight: uint64
+}
+```
+
+Comparison between heights is implemented as follows:
+
+```typescript
+function compare(a: Height, b: Height): Ord {
+  if (a.epochNumber < b.epochNumber)
+    return LT
+  else if (a.epochNumber === b.epochNumber)
+    if (a.epochHeight < b.epochHeight)
+      return LT
+    else if (a.epochHeight === b.epochHeight)
+      return EQ
+  return GT
+}
+```
+
+This is designed to allow the height to reset to `0` while the epoch number increases by one in order to preserve timeouts through zero-height upgrades.
+
+### Headers
+
+Contents of wasm client headers depend upon `WASM Client Code`.
+
+```typescript
+interface Header {
+  data: []byte
+}
+```
+
+### Misbehaviour
+
+The `Misbehaviour` type is used for detecting misbehaviour and freezing the client - to prevent further packet flow - if applicable.
+WASM client `Misbehaviour` consists of two headers at the same height both of which the light client would have considered valid.
+
+```typescript
+interface Misbehaviour {
+  fromHeight: Height
+  h1: Header
+  h2: Header
+}
+```
+
+### Client initialisation
+
+WASM client initialisation requires a (subjectively chosen) latest consensus state interpretable by the target WASM Client Code. `wasmCodeId` field is unique identifier for `WASM Client Code`, and `initializationData` refers to opaque data required for initialization of the particular client managed by that WASM Client Code.
+
+```typescript
+function initialise(
+    wasmCodeId: String,
+    initializationData: []byte,
+    consensusState: []byte,
+  ): ClientState {
+    codeHandle = getWASMCode(wasmCodeId)
+    assert(codeHandle.isInitializationDataValid(initializationData, consensusState))
+    set("clients/{identifier}/consensusStates/{height}", consensusState)
+    return codeHandle.initialise(initializationData, consensusState)
+}
+```
+
+The `latestClientHeight` function returns the latest stored height, which is updated every time a new (more recent) header is validated.
+
+```typescript
+function latestClientHeight(clientState: ClientState): Height {
+  codeHandle = clientState.codeHandle();
+  codeHandle.latestClientHeight(clientState)
+}
+```
+
+### Validity predicate
+
+WASM client validity checking uses underlying WASM Client code. If the provided header is valid, the client state is updated & the newly verified commitment written to the store.
+
+```typescript
+function checkValidityAndUpdateState(
+  clientState: ClientState,
+  epoch: uint64,
+  header: Header) {
+    codeHandle = clientState.codeHandle()
+    isValid, consensusState = codeHandle.validateHeaderAndCreateConsensusState(clientState, header, epoch)
+    set("clients/{identifier}/consensusStates/{header.height}", consensusState)
```

Comment:
can we give the WASM code limited state read/write access instead, it would be more flexible

Reply:
I agree completely. Since you are using wasmvm, it would not be hard to give each contract a sandboxed storage area, like we do in `x/wasm`. The contract knows what data it needs and can just store it locally.

However, it seems you actually need to write to some particular ibc keys here. Could you not just make a prefix store of `clients/{identifier}` and pass that to the contract. It could then store client state under the empty key and  consensus state under `/consensusStates/{header.height}`. Meaning full access to the state it manages, but no access to other ibc state

Reply:
@ethanfrey @cwgoes I did not want WASM code read/write  access to storage, because current version of WASM manager does not allow validation of the WASM code thoroughly. So, to err on safe side, I removed storage access to the WASM code.

Reply:
If wasm can return results that will be written to some specified storage locations, you could also provide the wasm code direct read/write access to only those storage locations.

I would assume wasm code is potentially malicious (even if this code must have been verified by governance to run), but you should be able to define a sandbox where it can work without impacting other modules (just it's own state). It is trusted to approve packets for a subset of clients/connections/channels, so should be able to write any data relative to those without increasing the attack surface at all

Reply:
What do you mean, "does not allow validation of the WASM code thoroughly"? That sounds dangerous. I would be far more worried about compute issues than storage issues, since storage usage is gas-metered and safely sub-store-constrained by the keeper.

Reply:
I think @ParthDesai is referring to [this piece of code](https://github.com/ChorusOne/ibc-go/blob/ics_28_wasm_client/modules/core/28-wasm/keeper/validation.go#L22)

Right now, Wasm manager validates by instantiating the Wasm VM (via CosmWasm lib). There is no code introspection (unless provided by CosmWasm)

Reply:
> Right now, Wasm manager validates by instantiating the Wasm VM (via CosmWasm lib). There is no code introspection (unless provided by CosmWasm)

Sure. But who can define such a wasm code as a light client? This is out of scope for ICS repo maybe, but I feel essential to understand how this integrates in a blockchain, and what security guarantees we need. 

Can I upload some wasm as `light-client-foo` and then make a new client/connection on top of it? Does governance need to approve it? If this code misbehaves what is the worst that can happen? (I assume fake packets in any channels built on top of this client).

Reply:
> If wasm can return results that will be written to some specified storage locations, you could also provide the wasm code direct read/write access to only those storage locations.

Agreed. The WASM code handlers should just be given the client-id prefixed store. This is **already** what is given to the client handlers. Core IBC already prevents the ability for one client to read/write outside of its sandbox. So WASM can follow same security model. Just pass the store that the light client receives directly into WASM, it is already sandboxed by core IBC.

> But who can define such a wasm code as a light client? This is out of scope for ICS repo maybe, but I feel essential to understand how this integrates in a blockchain, and what security guarantees we need.

> Can I upload some wasm as light-client-foo and then make a new client/connection on top of it? Does governance need to approve it? If this code misbehaves what is the worst that can happen? (I assume fake packets in any channels built on top of this client).

We may rollout more conservatively, but I'm of the belief that the spec itself should enable random users uploading wasm light client code. This is a big selling point of WASM clients imo, since you don't need buy in from governance to support your eclectic choice of consensus algorithm.

Since IBC is designed to sandbox any issues, any malicious light client code submitted by user can only affect users of that light client. Though of course, they can do whatever they want to the users of their light client.

Here it will be incumbent on counterparty chains to enforce wasm contract correctness, by saying I will only accept a connection where the client has this smart contract code hash(es). This way we do not allow malicious clients to be submitted against correct chains. By creating the restriction on the destination chain rather than the source chain, we can enable all the flexibility of dynamic light clients while still preventing malicious connections over correctly implemented chains.

Reply:
Though we should double check that `02-client` does not make any assumptions that something will be in the store in ways that will affect its overall functioning.

It's ok if the client becomes non-functional or downright malicious imo.

What would be bad, is if deleting something (like say the clientstate itself), causes the client handler to panic and shutdown the node. In case, the wasm contract does something like this, we should just return an error and fail th tx whenever client state is needed

Reply:
> Just pass the store that the light client receives directly into WASM, it is already sandboxed by core IBC.

In the [initial implementation](https://github.com/cosmos/ibc-go/pull/208) we already do that, this is, use the default flow where `02-client/keeper/client.go` gets client-id prefixed store from [ClientStore()](https://github.com/cosmos/ibc-go/blob/2523fdbfd8fc14554d81dc92ce58b3a590964647/modules/core/02-client/keeper/keeper.go#L386), and passes it on, to the client handle.

@AdityaSripal @cwgoes I am not sure how to phrase it (could use your help here), but here's a try:
```
    store = getStore("clients/{identifier}")
    isValid, consensusState = codeHandle.validateHeaderAndCreateConsensusState(store, clientState, header, epoch)
    store.set("/consensusStates/{header.height}", consensusState)
    // save the client
    store.set("", clientState)
```

> We may rollout more conservatively, but I'm of the belief that the spec itself should enable random users uploading wasm light client code. This is a big selling point of WASM clients imo, since you don't need buy in from governance to support your eclectic choice of consensus algorithm.

Agreed.

Reply:
To let the light-clients update their state, IBC interface would have to change slightly (it's going to be covered by https://github.com/cosmos/ibc/issues/569).

For example:
```
CheckHeaderAndUpdateState(sdk.Context, codec.BinaryCodec, sdk.KVStore, Header) (ClientState, ConsensusState, error)
```

to smth like this:
```
CheckHeaderAndUpdateState(sdk.Context, codec.BinaryCodec, sdk.KVStore, Header) error
```

because the {client,consensus} states are going to be updated within `CheckHeaderAndUpdateState` lightclient method